### DISCUSSION TITLE
Random selection between matching rules based on their weights.

### DISCUSSION THREAD

**Discussion by Inferdy** [mxgmn/MarkovJunior#37] [Ideas]
Let's say we have a set of rules with their weights:
```
{
  {A, 5},
  {B, 2},
  {C, 2},
  {D, 1}
}
```
And rules A, C and D are active (their inputs have atleast 1 match on the grid). It means we have to apply one of these rules based on their weights. The sum of their weights is `5 + 2 + 1 = 8`, so the chance of rule A to be applied is `5/8`, the chance of rule C to be applied is `2/8`, D - `1/8`.

It will be useful for selecting different possible building parts. For example, we could build a dungeon placing tetris pieces adjacently. Not all the pieces can be placed on some states of the grid and we want to make them have different weights to set more preferrable  pieces.

**Comment by mxgmn** [comment]
Adding weights is in the plans, yes.

**Reply by Inferdy** [reply]
Random selection between matching rules based on matches count would be cool, too.

**Reply by mxgmn** [reply]
What do you mean, based on matches count? You can emulate your original example like this:
```xml
<all>
  <rule in="W" out="A"/>
  <rule in="W" out="A"/>
  <rule in="W" out="A"/>
  <rule in="W" out="A"/>
  <rule in="W" out="A"/>
  <rule in="W" out="B"/>
  <rule in="W" out="B"/>
  <rule in="W" out="C"/>
  <rule in="W" out="C"/>
  <rule in="W" out="D"/>
</all>
```

**Reply by Inferdy** [reply]
I think
```xml
<one>
   <rule in="W" out="R"/>
   <rule in="W" out="G"/>
   <rule in="W" out="B"/>
</one>
```
Is just what I need. If this construction makes sense 😂

**Reply by Inferdy** [reply]
I have a similar idea. We could choose different nodes based on their weights in case of a match.

Here's an example:
I have a dungeon-building algorithm which cuts 4-cell pieces (tetris pieces; only if possible). I wish I could make it prefer to cut 5 cells with a chance of 10%, cut 4 cells with a chance of 60%, cut 3 cells with a chance of 20% and cut 2 cells with a chance of 10%.
I could achieve this if I could choose 3 different nodes based on their weights after I successfuly destroyed 1 piece.

![6k0wap](https://user-images.githubusercontent.com/24939153/174361882-dc700a4c-0d36-4e7f-8cce-618a5e2dbfce.gif)

![image](https://user-images.githubusercontent.com/24939153/174358017-f752bec6-c90a-4822-9ae6-a045a722f189.png)

**Reply by Inferdy** [reply]
Well, it feels like we need scripting on top of the nodes interfaces.

Nodes like `one`, `all`, `prl` are useful on their own, but our interaction with them is limited by markov orders or sequence orders for some reason. We can not customize the steps count and build complex scenarios.

**Reply by Inferdy** [reply]
Ah, seems like we are limited in our interactions with rulenodes to make probabilistic inference easier 🤔

**Reply by mxgmn** [reply]
Could you draw the final result that you want to achieve? Maybe it's possible to do with just Markov nodes.
